x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobDataCountsPersister.java

```diff
@@ -33,30 +35,60 @@
 
     private static final Logger logger = LogManager.getLogger(JobDataCountsPersister.class);
 
+    private final ResultsPersisterService resultsPersisterService;
     private final Client client;
+    private final AnomalyDetectionAuditor auditor;
 
-    public JobDataCountsPersister(Client client) {
+    public JobDataCountsPersister(Client client, ResultsPersisterService resultsPersisterService, AnomalyDetectionAuditor auditor) {
+        this.resultsPersisterService = resultsPersisterService;
         this.client = client;
+        this.auditor = auditor;
     }
 
-    private XContentBuilder serialiseCounts(DataCounts counts) throws IOException {
+    private static XContentBuilder serialiseCounts(DataCounts counts) throws IOException {
         XContentBuilder builder = jsonBuilder();
         return counts.toXContent(builder, ToXContent.EMPTY_PARAMS);
     }
 
     /**
      * Update the job's data counts stats and figures.
+     * NOTE: This call is synchronous and pauses the calling thread.
+     * @param jobId Job to update
+     * @param counts The counts
+     */
+    public void persistDataCounts(String jobId, DataCounts counts) {
+        try {
+            resultsPersisterService.indexWithRetry(jobId,
+                AnomalyDetectorsIndex.resultsWriteAlias(jobId),
+                counts,
+                ToXContent.EMPTY_PARAMS,
+                WriteRequest.RefreshPolicy.NONE,
+                DataCounts.documentId(jobId),
+                () -> true,
+                (msg) -> auditor.warning(jobId, "Job data_counts " + msg));
+        } catch (IOException ioe) {
+            logger.warn(() -> new ParameterizedMessage("[{}] Error serialising DataCounts stats", jobId), ioe);
```

Comment:
if its an error, should it be `logger.error`? I am not sure what you mean with "serialising error"

The error handling here silences any problem, I do not think we should do this, but should only be forgiving regarding failed index requests.

Reply:
🤔 

Possibly, but failing the whole job because something weird happened writing data counts out seems like overkill. 

Yeah, I will change these to `logger.error` for sure and tighten up the log line.